ui/components/Shared/SharedInput.tsx

```diff
@@ -49,6 +59,22 @@ export default function SharedInput(props: Props): ReactElement {
     }
   })
 
+  const onInputChange: ChangeEventHandler<HTMLInputElement> = ({
+    target: { value: inputValue },
+  }) => {
+    if (parser) {
+      const result = parser(inputValue)
+      if (result.state === "error") {
+        setParserError(result.message)
+      } else {
+        setParserError(null)
+        onChange?.(result.parsed)
+      }
+    } else {
+      onChange?.(value as unknown as T)
+    }
+  }
```

Comment:
What do you think about making the split `{ error: string } | { parsed: T }`? This feels less noisy in the parser function, but still gives us a good way to discriminate at the top level. It's also clear in the handling code what path is doing what:

```suggestion
  const onInputChange: ChangeEventHandler<HTMLInputElement> = ({
    target: { value: inputValue },
  }) => {
    if (parser) {
      const result = parser(inputValue)
      if ("error" in result) {
        setParserError(result.error)
      } else {
        setParserError(null)
        onChange?.(result.parsed)
      }
    } else {
      onChange?.(value as unknown as T)
    }
  }
```

Reply:
Oh nice, the `"error" in result` gives TypeScript enough info to know `result.parsed` exists as T in the `else`. I did not consider that. I dig it. We can still add a `{ warning: string }` to the union if we want to expand out 👍 

✔️